Guide: PSTs #90

felwintrr opened the pull request:
Guide for deploying a PST.

Wondering if I should have included the contract code for the template contract I linked to and referenced.

Ofc any other suggestions welcome too! :)

Thanks guys

twilson63 commented:
I wonder if we should still show smart weave it is kind of a deprecated module. Maybe better to show how to deploy a PST using Warp to take advantage of the performance and L2 efficiency?

@felwintrr @DanMacDonald - Thoughts?

felwintrr commented:
Super good point - I actually didn't know it was depreciated 😓 

I'm happy to go with whatever you guys think is best. If Warp is the standard, it might be best for me to go and explore Warp and come back with a more up-to-date guide.

twilson63 commented:
With Warp they have specific pst apis

```
import { WarpFactory } from 'warp-contracts'

const warp = WarpFactory.forMainnet()

// deploy contract
await warp.createContract.deploy({
  wallet,
  initState,
  src
})

// pst apis
const pst = warp.pst(CONTRACT)

const state = await pst.currentState()

// transfer
await pst.transfer({target, qty})

DanMacDonald commented:
Yeah, better to go with Warp at this point. smartweave.js only handles L1 transactions, which means the state of the contract is reliable as of about 8 confirmations, anything newer than that can lead to diverging states locally. It will run into scale problems in the near future as well, with the Arweave Block transaction limit being capped at 1000.

twilson63 commented:
Here is the Warp CLI, I would recommend using it to demonstrate how to deploy contracts - https://www.npmjs.com/package/warp-contracts-cli

The automatic use of the sequencer and bundlr to deploy contracts creates a quality and consistency that is significantly a better result than the legacy smartweave CLI

felwintrr commented:
thanks guys, let me update this to warp.

felwintrr commented:
Have rewrote this, but some points I'd love to hear your feedback/thoughts on:

- Is there too much crossover about smart weave contracts as a whole? Should I thin that bit down a bit?
- In the Warp PST docs, there is an input function `mint` which lets the caller mint new tokens. I changed this and added a conditional to check the caller is the owner first - not sure if I should do this? But unless I'm misunderstanding it means anyone who calls the contract can mint new tokens
- Is there a need for `async` in the `handle` method signature? Warp seems to use it, but can't see it used anywhere else
- Still need to to a guide on calling PST in frontend? 

And of course any other thoughts you have feel free to share😄

Thanks guys

DanMacDonald commented:
![image](https://user-images.githubusercontent.com/3269261/214195255-406e65cc-cc3a-4831-8d7c-245137ee4461.png)
I'm not sure this makes sense as a best practice. The balances are readily available in the contract state so the preferred way would just be to `readContractState()` on the contract and look up the balance, as opposed to posting a transaction to return the balance as this transaction really only serves to slow down the evaluation of the contract over time.

felwintrr commented:
okay thanks @DanMacDonald - I was mainly reading from [here](https://academy.warp.cc/tutorials/pst/writing-pst-contract/contract-source) and following along. 

I did also wonder if what I was doing was outdated, because of @twilson63's prev. example using the PST api. 

Is there a specific PST standard currently in place I could look at? I know there was the original guide or example in official Arweave docs, and in the Warp docs I linked it's fairly similar. I feel like there should be some kind of OZ/Solmate equivalent if there isn't already haha.

In the meantime, I'll keep digging into Warp and play around with both the PST api and the smartweave global api and see if I can make this any better.

twilson63 commented:
@felwintrr @DanMacDonald Everywhere I have seen, they include the function of the balance, but I don't think an Action is written to Arweave to calculate a read-only function. (I could be wrong), When I looked at this before, the Warp SDK, evaluator gets the current state and evaluates the read-only function against the current state. 

Whether to include or not I think, is up to you; I guess it could be useful, if you needed to format some state consistently when reading the values, like an aggregate or a filter or something.

twilson63 reviewed (commented):
In your transfer contract you need to check

`if (balances[caller] < qty) {` 

balances[caller] !== undefined

If you don't make this check, then someone could basically mint tokens to their address via the transfer method.

You may consider just adding this code to the top of your handle function

```js
if (balances[caller] === void 0 || balances[caller] === null) {
    balances[caller] = 0;
  }  
  
 ```

DanMacDonald commented:
Interesting @twilson63 , so the `readState()` doesn't actually post an interaction TX,  but `interactWrite()` does? That would make sense if it worked that way... 🤔
It's kind of disappointing that there isn't a way to enforce that data access pattern, it appears that people COULD post interactions for these check balance calls which is less than ideal.

twilson63 reviewed (approved):
lgtm
